### ISSUE TITLE
is_file check on html

### ISSUE THREAD

**Issue by blaaat** [mikehaertl/phpwkhtmltopdf#307]
PHP Warning is_file(): open_basedir restriction in effect. File(&lt;!DOCTYPE HTML&gt;
&lt;html&gt;

phpwkhtmltopdf/src/Pdf.php:325 mikehaertl\wkhtmlto\Pdf::processOptions

On my production machine (CentOS7; PHP 7.2) PHP_MAXPATHLEN is quite high. Somehow the html header-footer-html triggers the openbase_dir error in the is_file call. 

Maybe better to do the other checks (html/xml/url regex or strip_tags) before the is_file call?

**Comment by blaaat** [comment]
For now; if anyone is interested; this is my work-around; 

```
class Pdf extends \mikehaertl\wkhtmlto\Pdf
{
	public function setHeader($input)
	{
		return $this->setOptionHtml('header-html', $input);
	}
	public function setFooter($input)
	{
		return $this->setOptionHtml('footer-html', $input);
	}
	
	protected function setOptionHtml($option, $input) {
        $options[$option] = $this->_tmpFiles[] = new File($input, '.html', self::TMP_PREFIX, $this->tmpDir);
        $this->setOptions($options);
	}
}
```

**Comment by mikehaertl** [comment]
> Maybe better to do the other checks (html/xml/url regex or strip_tags) before the is_file call?

It's tricky. We simply can't know, that what you passed is some static text or a file name. Using `is_file()` was the only option we saw. Note, that what you pass here doesn't need to be HTML. If you see a better way, feel free to suggest a change. Also see #100 where we already changed this.

**Comment by mikehaertl** [comment]
We could maybe add another check for HTML via the Regex we already have. This would at least help in the cases, where a full HTML document is passed.

**Comment by blaaat** [comment]
Isn't the `processInput()` method suitable to reuse?

**Comment by mikehaertl** [comment]
Yeah, your're probably right. It's been a while since I wrote this. I'll try to modify things a little and find better method names. Will let you know when I've updated it. It would be great if you could then help testing.

**Comment by blaaat** [comment]
Thanks! and off course!

**Comment by mikehaertl** [comment]
@blaaat I've created a MR with my refactorings, see the comments there. Could you give it a try?

As  a side note: This changes the default behavior. Since we follow Semver this change will probably also require another major release.

**Comment by blaaat** [comment]
Thanks, I'll and let you know. 

The `strip_tags` way to determine HTML looks risky. A path might include a tag which could be stripped:

> /home/files<2018>/test.html


I'd prefer the old regex (maybe even with a ^prefixed)

**Comment by blaaat** [comment]
> @blaaat I've created a MR with my refactorings, see the comments there. Could you give it a try?

Fixes my problem! Thanks.

**Comment by mikehaertl** [comment]
> The strip_tags way to determine HTML looks risky. A path might include a tag which could be stripped:

Good point, thanks. I somehow assumed `<` and `>` are not valid filename characters - but they are. So I've reverted to the old regex HTML check and removed the `strip_tags()` completely. I've not added the `^` though, as this would definitely break BC: Before it also accepted strings with content before `<html>`.

Right now it tries to stay away from `is_file()` as long as possible. There may still be situations where you hit your initial problem. So I've now also included the option to pass a `File` instance if all other methods fail.

```
$pdf->setOption([
    'header-html' => new File('some complex content'),
]);
```

See the updated MR here: https://github.com/mikehaertl/phpwkhtmltopdf/pull/308

Maybe you can take another look?

**Comment by mikehaertl** [comment]
Great. Just released 2.4.0 including this change. Thanks for your help!
